### DISCUSSION TITLE
Sample usage of notifications: detect control width change

### DISCUSSION THREAD

**Discussion by RablinQ** [jovibor/HexCtrl#96] [Ideas]
I have a dialog containing the control just like in the provided samples.
What I want is that when the size of the control changes (due to font change, capacity change or group view change) the parent dialog automatically resizes to the correct width so the control fits horizontally.

Because all those internal changes are notified using WM_NOTIFY and we can get the actual width of the control, this resizing can be implemented.

I had to find out how to do that. There seems no implementation for WM_NOTIFY in the sample.
Based on the documentation I had to find out how to handle WM_NOTIFY in MFC myself.

I would like to share my solution here for others. But also have a feature request to make it more elegant:

The 3 messages below are enough to implement the auto resize:

```cpp
HEXCTRL_MSG_SETFONT
HEXCTRL_MSG_SETGROUPSIZE
HEXCTRL_MSG_SETCAPACITY
```

But it might be a good idea to introduce a new notification just to inform the width of the control changed due to any reason.
My resize feature is not interested in why the width changed. I only need a notification that it occurred.

For example:

HEXCTRL_MSG_WIDTHCHANGED

Ok, and now my solution. I used the **Sample Dialog** project provided with HexCtrl.

First you have to add the 3 messages ON_NOTIFY to the message map of the parent dialog.
In SampleDialogDlg.cpp:

```cpp
BEGIN_MESSAGE_MAP(CSampleDialogDlg, CDialogEx)
	ON_WM_CLOSE()
	ON_BN_CLICKED(IDC_SETDATARND, &CSampleDialogDlg::OnBnSetRndData)
	ON_BN_CLICKED(IDC_CLEARDATA, &CSampleDialogDlg::OnBnClearData)
	ON_BN_CLICKED(IDC_FILEOPEN, &CSampleDialogDlg::OnBnFileOpen)
	ON_BN_CLICKED(IDC_HEXPOPUP, &CSampleDialogDlg::OnBnPopup)
	ON_BN_CLICKED(IDC_CHK_RW, &CSampleDialogDlg::OnChkRW)
	ON_WM_DROPFILES()
	ON_WM_PAINT()
	ON_WM_QUERYDRAGICON()
	ON_WM_SIZE()
	ON_NOTIFY(HEXCTRL_MSG_SETFONT, IDC_MY_HEX, &CSampleDialogDlg::OnWidthChanged)
	ON_NOTIFY(HEXCTRL_MSG_SETGROUPSIZE, IDC_MY_HEX, &CSampleDialogDlg::OnWidthChanged)
	ON_NOTIFY(HEXCTRL_MSG_SETCAPACITY, IDC_MY_HEX, &CSampleDialogDlg::OnWidthChanged)
END_MESSAGE_MAP()
```

Then add the declaration of the **OnWidthChanged** method:
In SampleDialogDlg.h, class CSampleDialogDlg :

```cpp
afx_msg void OnWidthChanged(NMHDR* pNMHDR, LRESULT* pResult);
```

And lastly implement the action resizing code.
In SampleDialogDlg.cpp, 

```cpp
void CSampleDialogDlg::OnWidthChanged(NMHDR* pNMHDR, LRESULT* pResult)
{
	// We know the exact pixel width of the control using HexCtrl API
	int controlWidth = m_pHexDlg->GetActualWidth();
	// Retrieve the system metrics for a vertical scrollbar width when displayed. We need to make room for that
	int verticalScrollbarWidth = ::GetSystemMetrics(SM_CXVSCROLL);

	// Add the vertical scrollbar width to the control width
	controlWidth += verticalScrollbarWidth;

	//Well, maybe a few pixels short. Add a reasonable margin (hardcoded? Maybe a better way?)
	controlWidth += 10;

	// Get the current styles of the dialog
	DWORD dwStyle = GetStyle();
	DWORD dwExStyle = GetExStyle();

	// Calculate the required window size to ensure the client area fits the control width
	CRect adjustedRect(0, 0, controlWidth, 0); // Only care about width
	AdjustWindowRectEx(&adjustedRect, dwStyle, FALSE, dwExStyle);

	// Get the current dialog's position and height
	CRect windowRect;
	GetWindowRect(&windowRect);

	// Set the new width while keeping the current height and position
	int newWidth = adjustedRect.Width(); // Adjusted width with non-client area accounted for
	SetWindowPos(nullptr, windowRect.left, windowRect.top, newWidth, windowRect.Height(), SWP_NOZORDER | SWP_NOMOVE);

	*pResult = 0;
}
```

Now when using the scroll wheel on the control which changes the font size, the dialog automatically resizes to the width of the control so no horizontal scrollbar is shown. Any change to the width will trigger width resize.

When the dialog is created it automatically resizes the width to fit the control already as the font is set on the control which triggers the resize.

When this seems to be a nice addition to the sample maybe add this feature to the project.

**Comment by jovibor** [comment]
>There seems no implementation for WM_NOTIFY in the sample.
Based on the documentation I had to find out how to handle WM_NOTIFY in MFC myself.

`WM_NOTIFY` is a standard Windows message, it's MFC unrelated. You can handle it whichever way you like.

>HEXCTRL_MSG_WIDTHCHANGED

Basically what you want is to shift what you've just implemented, in your `OnWidthChanged` method, to the **HexCtrl** shoulders.
While it's of course possible, the one slight downside I can see is that parent window will now always receive two notification messages simultaneously, `HEXCTRL_MSG_WIDTHCHANGED` and for example `HEXCTRL_MSG_SETFONT`.
It's not a big deal per se, but have to think about such design.

**Reply by RablinQ** [reply]
> Basically what you want is to shift what you've just implemented, in your `OnWidthChanged` method, to the **HexCtrl** shoulders. 

Well. My sample could be added to the HexCtrl sample (or maybe only show the use of HEXCTRL_MSG_SETFONT how to implement notifications. I did not suggest to add the automatic resizing into the HexCtrl itself. I think that is the responsibility of the developer using HexCtrl. The suggestion is to add HEXCTRL_MSG_WIDTHCHANGED to the API of HexCtrl so that clients can react to it without "misusing" the 3 notifications it offers now. Maybe in future there could be other actions within the control which resiults in a width change so then developers need to add those new events to keep the resizing working. The HEXCTRL_MSG_WIDTHCHANGED encapsulates any action within the control, now and in future, with one single notification.

**Reply by jovibor** [reply]
I must have phrased it incorrectly. 
I meant how you combined all three notifications in one `OnWidthChanged` callback method. Not the implementation itself of course.

**Reply by jovibor** [reply]
>My sample could be added to the HexCtrl sample (or maybe only show the use of HEXCTRL_MSG_SETFONT how to implement notifications.

By the way, if you'd like to add any improvement to the HexCtrl you think would be helpful for others, you can always submit a PR.
